batch2/batch/driver/scheduler.py

```diff
@@ -19,6 +20,33 @@ def __init__(self, app):
         asyncio.ensure_future(self.loop('cancel_loop', self.cancel_state_changed, self.cancel_1))
         asyncio.ensure_future(self.bump_loop())
 
+    async def compute_fair_share(self):
+        free_cores_mcpu = sum([worker._free_cores_mcpu for worker in self.inst_pool.healthy_instances_by_free_cores])
+        total_cores_mcpu = len(self.inst_pool.healthy_instances_by_free_cores) * self.inst_pool.worker_cores * 1000
+        allocated_cores = {}
+
+        records = self.db.execute_and_fetchall(
+            '''
+            SELECT user, ready_cores_mcpu, running_cores_mcpu
+            FROM user_resources
+            WHERE ready_cores_mcpu > 0
+            ORDER BY ready_cores_mcpu ASC;
+            ''')
+
+        n_users = len(records)
+        n_unallocated_users = n_users
+        async for record in records:
```

Comment:
OK, this is better, but there's a subtly here.  Are we dividing up the free cores or the total cores?  You're dividing the free cores with a cap at an even share of the total cores.  I think we should be dividing up the total cores, e.g. if two users are running on 1000 cores, our share is each 500.

Once you compute the total target share, you can assign free cores to users.  A few ways you can do this:
 - The user's need is max(0, allocated - running), give the each user their fraction of the total need * free cores
 - Allocate their total share, and just schedule what you can in the scheduler.  This is clearly the easiest.
 - Allocate their total share, and overschedule the cluster.

Reply:
What I was hoping this would do was give each user at a minimum up to their fair share. If there's extra cores from the preceding users not needing their fair share, I wanted those shares to get allocated to the next user if they need it. So what I think I want is actually this:

```
max(0, min(record['ready_cores_mcpu'],
                    (total_cores_mcpu / n_users) - record['running_cores_mcpu'] + free_cores_mcpu / n_unallocated_users))
```